Fine grained error handling for Salt state

slaperche-scality opened the pull request:
**Component**:

operator, test, salt

**Context**:

Make the distinction between Salt state error resulting from the execution of the state itself and the others.

**Summary**:

Check `retcode`.
Add some new tests (and fix bugs uncovered by those).

**Acceptance criteria**: 

The test suite is passing.

---

Closes: #1593

slaperche-scality committed:
storage-operator/salt: discriminate on retcode

We now make the disctinction between transient error (retcode 1 it
seems) versus "definitive" errors from our state (retcode 2 is seems…)

That way, we won't stay stuck retrying forever a Salt state that fails.
Instead we will move the Volume into a Failed state (which allows it to
be deleted).

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: code simplification

Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: add test on creation failure

Now that we correctly handle the failure of our Salt state, we can add
tests for such cases.

Sadly, we can't have a good error message for when the Volume is not in
the pillar, because it seems that you can't have a customized comment
for `test.fail_without_changes` due to a bug in our version of Salt (see
https://github.com/saltstack/salt/issues/51821)

Refs: #1417
Closes: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: extract root cause from state error

When a Salt state fail, we try to find the root cause of the failure
(i.e. a state that has failed and not because of its requisite).

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/volume: ignore failed JID

When we're trying to delete a job in a failed state, it may already have
a Job ID (if the Volume failed during its creation) that we kept around
for troubleshooting purpose (one may want to manually query Salt API
with the Job ID to get more info about why it failed).

Now, we must ignore this Job ID otherwise we will wrongly think that the
UnprepareVolume is already in progress…
So let's ignore Job ID from Failed Volume, once we start to unprepare
the volume we will get our own Job ID and move into the Terminating
state (so that the Job ID won't be ignored in the next iterations).

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: fix unprepared state

If a volume is not found in the pillar, we shouldn't fail. Instead we
should consider that we have nothing to do.

Indeed, consider what happens if you try to create a Volume which is not
in the pillar (because it uses a non-existing storage class for
instance): this Volume will fail with a CreationError and when you'll
try to delete this Volume you've nothing to unprepare since it was never
created to begin with.
In such case, you don't want to have unprepare failing.

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

NicolasT commented on salt/metalk8s/volumes/unprepared/init.sls:
```diff
@@ -8,5 +8,5 @@ Clean up backing storage for {{ volume }}:
 {%- else %}
 Volume "{{ volume }}" not found in pillar:
-  test.fail_without_changes: []
+  test.succeed_without_changes: []
```
Is there some way to still render some kind of 'warning' message in the state's output, e.g. using `test.configurable_test_state` and setting the `comment` to something descriptive?

While indeed this should be 'success', it's also something that should not occur all too frequently, and something we may want to report to a user (beyond the `name` of the state).

TeddyAndrieux replied:
For me you should log warning in jinja so that you will have real warning output in salt log and in output https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.logmod.html
Because put something directly in the comment of the state result is not really usefull if it's success nobody will read it IMO.

NicolasT replied:
Then do both. Keep in mind that, when the operator invokes Salt, no human will see any Salt logs (except when explicitly looking at the logs on the minion system at that point in time).

slaperche-scality replied:
Done

NicolasT commented on storage-operator/pkg/salt/client.go:
```diff
@@ -198,11 +199,38 @@ func (self *Client) PollJob(
-	if int(retcode) != 0 {
+	switch int(retcode) {
+	case 0:
+		jobLogger.Info("Salt job succeedeed")
```
```suggestion
		jobLogger.Info("Salt job succeeded")
```

slaperche-scality replied:
Done

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

slaperche-scality committed:
salt/volume: use `configurable_test_state` to have better message

Use `test.configurable_test_state` instead of
`test.fail_without_changes` in order to provides a custom `comment` that
will give more context to the user (instead of simply `Failure!`).

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: don't filter out volume from pillar

We used to filter out the Volume without an associated StorageClass that
exists.
The problem is that if you delete a StorageClass after creating a volume
it should be perfectly OK (because we only use it at creation time), but
that's not true: even though the `unprepared` state doesn't use the
StorageClass, it does need the Volume. But since the Volume reference a
non-existing StorageClass, it's not present in the pillar and thus
cannot be cleaned up by the `unprepared` state.

To avoid this kind of situation we no longer filter out Volume
referencing a non-existing storage class. Instead, we store the Volume
in the pillar, but without the StorageClass information.
Then, the `prepared` state will check if we do have StorageClass
information, and if not it will raise an error.
Since `unprepared` doesn't care about the StorageClass, nothing will
happen there and we will happily clean up the Volume (because this time,
we will have an entry in the pillar).

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: test that we can delete a volume with missing SC

We rely on StorageClass only at creation time, after that they can be
deleted with no ill-effect on the already created volume.
So let's have a test about that:
- create a volume using a StorageClass SC
- delete SC
- try to delete the volume: it should work.

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: add StorageClass cleanup

Add the possibility to specify a prefix to select which objects you want
to delete (indeed, for StorageClass, we don't want to delete the
`metalk8s-prometheus` one, thus we can't simly delete ALL the
StorageClass).

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: get rid of unused host fixture parameter

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: check backing storage

In order to be sure that the Salt states are working correctly, we had
check on the backing storage.

Note that to check that the backing storage is deleted, we need the
Volume UUID but we can't get it as the Volume as already been deleted at
that point. To solve this problem, we add a context where the functions
that fetch the volume can store the volume info which then can be reused
by later steps.

`_quantity_to_bytes` is copied from the Salt module (really not fond of
code duplication, but dunno how to share code between test suite and
Salt tree…).

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/salt: allow to test for a list of modules

Update the step that check the perms to allow the use of a list.

Refs: #1593
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/salt: add a test for storage-operator auth

Add a test to ensure that the storage-operator can authenticate against
Salt API using its service account.

Refs: #1593, #1412
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/salt: add folds

Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

jbertran commented on storage-operator/pkg/salt/client.go:
```diff
@@ -206,11 +207,32 @@ func (self *Client) PollJob(
+	case string:
+		return error
+	case map[string]interface{}:
+		for state := range error {
```
Oh boy that's confusing... Can we have `state, _` or does Go not like that?

slaperche-scality replied:
renamed the variable to be more explicit

jbertran commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -226,7 +226,10 @@ def format(self, force=False):
                             .format(self.path))
-        params = self.get('spec.storageClassName.parameters')
+        storage_class = self.get('spec.storageClassName')
+        if isinstance(storage_class, basestring):
```
Much confused here, could we have a comment explaining why the negative case is a `basestring`? It's not immediately obvious since we're departing from the usual `get` behaviour.

slaperche-scality replied:
Yeah, deserve a comment indeed.

jbertran commented on tests/post/steps/test_volume.py:
```diff
@@ -373,6 +379,34 @@ def _check_file_content():
+def check_storage_is_created(context, host, name):
+    volume = context.get(name)
+    assert volume is not None, 'volume {} not found in context'.format(name)
+    assert 'sparseLoopDevice' in volume['spec'], 'unsupported volume type'
```
Worth mentioning in the error message that it's unsupported _for this test_.

slaperche-scality replied:
Done

jbertran reviewed (commented):
Noice. Just a couple of minor comments.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

slaperche-scality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
